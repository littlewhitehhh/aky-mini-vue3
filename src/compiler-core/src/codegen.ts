export function generate(ast) {
  // let code = "";

  const context: any = createCodegenContext();

  const { push } = context;
  // code += "return";
  push("return ");
  const functionName = "render";
  const args = ["_cxt", "_cache"];
  const signature = args.join(",");

  push(`function ${functionName}(${signature}){`);
  // code += ` function ${functionName}(${signature}){`;

  console.log(ast);

  // code += `return`;
  push(`return `);
  genNode(ast.codegenNode, context);
  // code += "}";
  push("}");

  return {
    // code: `return function render(_cxt,_cache,$props,$setup,$data,,$options){
    //   return "hi"
    // }`,
    code: context.code,
  };
}
function genNode(node: any, context: any) {
  const { push } = context;
  // code += ` "${node.content}"`;
  push(`'${node.content}'`);
  // return code;
}
function createCodegenContext() {
  const context = {
    code: "",
    push(source) {
      context.code += source;
    },
  };
  return context;
}
