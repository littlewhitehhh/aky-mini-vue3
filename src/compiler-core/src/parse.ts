import { NodeTypes } from "./ast";

const enum TagType {
  Start,
  End,
}

export function baseParse(content: string) {
  //message
  const context = createParseContext(content); //{source:message}
  //重构
  return createRoot(parseChildren(context, [])); //
  // return {
  //   children: [
  //     {
  //       type: "interpolation",
  //       content: { type: "simple_expression", content: "message" },
  //     },
  //   ],
  // };

  // return createRoot([
  //   {
  //     type: "interpolation",
  //     content: { type: "simple_expression", content: "message" },
  //   },
  // ]);
}

function createParseContext(content): any {
  return {
    source: content,
  };
}
function createRoot(children) {
  return {
    children,
  };
}

function parseChildren(context, ancestors) {
  // ancestors 祖先
  const nodes: any = [];

  while (!isEnd(context, ancestors)) {
    let node;
    //解析插值{{}}
    if (context.source.startsWith("{{")) {
      node = parseInterpolation(context);
    } else if (context.source[0] === "<") {
      //element类型
      if (/[a-z]/.test(context.source[1])) {
        console.log("parse.element");
        node = parseElement(context, ancestors);
      }
    }
    // } else {
    //   //text类型
    //   node = parseText(context);
    // }
    if (!node) {
      node = parseText(context);
    }
    nodes.push(node);
  }
  return nodes;
  // return [
  //   {
  //     type: "interpolation",
  //     content: { type: "simple_expression", content: "message" },
  //   },
  // ];
}

function isEnd(context, ancestors) {
  //<div></div>

  //结束的条件
  //1、context.source 没有值的时候
  //2、当遇到结束标签时候

  //2

  const s = context.source;
  //</div>
  if (s.startsWith("</")) {
    for (let i = ancestors.length - 1; i >= 0; i--) {
      const tag = ancestors[i].tag;
      if (startsWithEndTagOpen(s, tag)) {
        return true;
      }
    }
  }

  // if (parentTag && s.startsWith(`</${parentTag}>`)) {
  //   return true;
  // }
  //1
  return !context.source;
}

//element类型parse
function parseElement(context: any, ancestors) {
  //implement    <div></div>
  // 1、解析tag
  // 2、删除处理完成的代码

  //1
  // const match: any = /^<([a-z]*)/i.exec(context.source); //
  // console.log(match); // [ '<div', 'div', index: 0, input: '<div></div>', groups: undefined ]
  // const tag = match[1];
  // console.log(tag); //div

  // //2  删除处理完成的代码
  // advanceBy(context, match[0].length);
  // console.log(context.source); //></div>
  // advanceBy(context, 1);
  // console.log(context.source); // </div>

  // return {
  //   type: NodeTypes.ELEMENT,
  //   tag,
  // };

  // 重构
  const element: any = parseTag(context, TagType.Start);
  ancestors.push(element);
  element.children = parseChildren(context, ancestors);
  ancestors.pop();

  // if (context.source.slice(2, 2 + element.tag.length) === element.tag) {

  //重构
  if (startsWithEndTagOpen(context.source, element.tag)) {
    parseTag(context, TagType.End);
  } else {
    throw new Error(`缺少结束标签:${element.tag}`);
  }

  console.log("-------", context.source);

  return element;
}

function parseTag(context: any, TagType) {
  //implement    <div></div>
  // 1、解析tag
  // 2、删除处理完成的代码

  //1
  const match: any = /^<\/?([a-z]*)/i.exec(context.source);
  console.log(match); // [ '<div', 'div', index: 0, input: '<div></div>', groups: undefined ]
  const tag = match[1];
  console.log(tag); //div

  // 2
  advanceBy(context, match[0].length); //></div>

  advanceBy(context, 1); //</div>

  if (tag === TagType.End) return;

  return {
    type: NodeTypes.ELEMENT,
    tag,
  };
}

//{{}}插值类型parse
function parseInterpolation(context) {
  console.log(context);

  const openDelimiter = "{{";
  const closeDelimiter = "}}";

  //{{message}}解析
  const closeIndex = context.source.indexOf(closeDelimiter, openDelimiter.length);

  advanceBy(context, openDelimiter.length);

  // context.source = context.source.slice(openDelimiter.length); // message}}
  console.log(context.source);

  const rawContentLength = closeIndex - openDelimiter.length;
  // const rawContent = context.source.slice(0, rawContentLength); // message
  const rawContent = parseTextData(context, rawContentLength);

  //边缘处理  利于 {{ message}}  双括号中有空格
  const content = rawContent.trim();
  console.log(content);

  // 例如{{message}}</div>   {{message}}处理完毕的 删除掉 剩下</div>
  // context.source = context.source.slice(closeDelimiter.length);
  advanceBy(context, closeDelimiter.length);
  console.log("context.source", context.source);

  return {
    type: NodeTypes.INTERPOLATION, //"interpolation"
    content: { type: NodeTypes.SIMPLE_EXPRESSION, content: content }, //"simple_expression"
  };
}

//TEXT类型parse
function parseText(context) {
  //1、获取content
  //2、推进代码  ： 删除已经处理的代码

  //1
  // const content = context.source.slice(0, context.source.length);
  // console.log(content);

  //2
  // advanceBy(context, content.length);
  // console.log(context.source);

  let endIndex = context.source.length;
  let endTokens = ["<", "{{"];

  for (let i = 0; i < endTokens.length; i++) {
    const index = context.source.indexOf(endTokens[i]);
    if (index !== -1 && endIndex > index) {
      //！==-1 证明没有插值
      endIndex = index;
    }
  }

  //重构
  const content = parseTextData(context, endIndex);
  console.log("content _______", content);

  return {
    type: NodeTypes.TEXT,
    content,
  };
}

//推进工具函数
function advanceBy(context: any, length: number) {
  context.source = context.source.slice(length);
}

//裁剪工具函数
function parseTextData(context, length) {
  const content = context.source.slice(0, length);

  advanceBy(context, length);
  return content;
}

function startsWithEndTagOpen(source, tag) {
  return source.startsWith("</") && source.slice(2, 2 + tag.length).toLowerCase() === tag.toLowerCase();
}
