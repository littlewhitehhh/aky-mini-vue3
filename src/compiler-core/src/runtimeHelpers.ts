export const To_Display_String = Symbol("toDisplayString");

export const helperMapName = {
  [To_Display_String]: "toDisplayString ",
};
