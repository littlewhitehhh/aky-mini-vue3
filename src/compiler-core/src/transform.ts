export function transform(root, options = {}) {
  const context = createTransformContext(root, options);

  //1、遍历 ——深度优先搜索  递归
  // 2、修改text content
  traverseNode(root, context);

  createRootChildren(root);
}

function createTransformContext(root: any, options: any) {
  const context = {
    root,
    nodeTransforms: options.nodeTransforms || [],
  };

  return context;
}

function traverseNode(node: any, context) {
  console.log(node);
  //1.element

  const nodeTransforms = context.nodeTransforms;
  for (let i = 0; i < nodeTransforms.length; i++) {
    const transform = nodeTransforms[i];
    transform(node);
  }

  traverseChildren(node, context);
}

function traverseChildren(node, context) {
  const children = node.children;

  if (children) {
    for (let i = 0; i < children.length; i++) {
      const node = children[i];
      traverseNode(node, context);
    }
  }
}
function createRootChildren(root: any) {
  root.codegenNode = root.children[0];
}
