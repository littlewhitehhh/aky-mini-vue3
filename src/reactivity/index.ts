export function add(a, b) {
  return a + b;
}

export { ref, proxyRefs } from "./ref";
