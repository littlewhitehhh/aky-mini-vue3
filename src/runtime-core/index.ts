// export { createApp } from "./createApp";

export { h } from "./h";

export { renderSlots } from "./helpers/renderSlots";

export { createTextVnode, createElementVnode } from "./vnode";

export { getCurrentInstance, registerRuntimeCompiler } from "./component";

export { provide, inject } from "./apiInject";

export { createRenderer } from "./renderer";

export { nextTick } from "./scheduler";
export { toDisplayString } from "../shared";

export * from "../reactivity/index";
