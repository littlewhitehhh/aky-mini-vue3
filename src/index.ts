// mini-vue 的出口

export * from "./runtime-dom/index";

export * from "./reactivity/index";

import { baseCompile } from "./compiler-core/src";
import * as runtimeDom from "./runtime-dom";
import { registerRuntimeCompiler } from "./runtime-dom";

// render

function compoileToFunction(template) {
  const { code } = baseCompile(template);
  const render = new Function("Vue", code)(runtimeDom);
  return render;
}
registerRuntimeCompiler(compoileToFunction);
