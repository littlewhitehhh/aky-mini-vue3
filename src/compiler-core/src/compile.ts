import { generate } from "./codegen";
import { baseParse } from "./parse";
import { transform } from "./transform";
import { transformElement } from "./transforms/tarnsformElement";
import { transformExpression } from "./transforms/transformExpression";
import { transformText } from "./transforms/transfromText";

export function baseCompile(template) {
  const ast: any = baseParse(template);
  transform(ast, {
    nodeTransforms: [transformExpression, transformElement, transformText],
  });
  console.log("ast_______", ast);

  return generate(ast);
}
