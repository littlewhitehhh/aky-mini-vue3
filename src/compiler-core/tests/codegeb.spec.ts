import { generate } from "../src/codegen";
import { baseParse } from "../src/parse";
import { transform } from "../src/transform";

describe("codegen", () => {
  it("string", () => {
    const ast = baseParse("hi");
    transform(ast);
    const { code } = generate(ast);

    //快照 作用
    //1、抓bug
    // 2、有意的改变
    expect(code).toMatchSnapshot();
  });
});
