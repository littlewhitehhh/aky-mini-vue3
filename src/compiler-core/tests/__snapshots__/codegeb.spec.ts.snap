// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen string 1`] = `"return function render(_cxt,_cache){return 'hi'}"`;
